File: src/main/java/no/nav/fo/veilarbregistrering/profilering/Innsatsgruppe.java
package no.nav.fo.veilarbregistrering.profilering;

import no.nav.fo.veilarbregistrering.besvarelse.Besvarelse;

import java.util.Arrays;

public enum Innsatsgruppe {
    STANDARD_INNSATS("IKVAL"),
    SITUASJONSBESTEMT_INNSATS("BFORM"),
    BEHOV_FOR_ARBEIDSEVNEVURDERING("BKART");

    private String arenakode;

    Innsatsgruppe(String arenakode) {
        this.arenakode = arenakode;
    }

    public String getArenakode() {
        return arenakode;
    }

    public static Innsatsgruppe of(String arenakode) {
        return Arrays.stream(Innsatsgruppe.values())
                .filter(innsatsgruppe -> innsatsgruppe.getArenakode().equals(arenakode))
                .findAny().orElse(null);
    }

    static Innsatsgruppe of(Besvarelse besvarelse, int alder, boolean harJobbetSammenhengendeSeksAvTolvSisteManeder) {
        if (besvarelse.anbefalerBehovForArbeidsevnevurdering()) {
            return Innsatsgruppe.BEHOV_FOR_ARBEIDSEVNEVURDERING;
        } else if (besvarelse.anbefalerStandardInnsats(alder, harJobbetSammenhengendeSeksAvTolvSisteManeder)) {
            return Innsatsgruppe.STANDARD_INNSATS;
        } else {
            return Innsatsgruppe.SITUASJONSBESTEMT_INNSATS;
        }
    }

}

File: src/main/kotlin/no/nav/fo/veilarbregistrering/profilering/Innsatsgruppe.kt
package no.nav.fo.veilarbregistrering.profilering

import no.nav.fo.veilarbregistrering.besvarelse.Besvarelse
import no.nav.fo.veilarbregistrering.metrics.Metric
import java.util.*

/**
 * https://confluence.adeo.no/display/KARTLEGGING/Begrep+og+forkortelser
 */
enum class Innsatsgruppe(val arenakode: String) : Metric {
    STANDARD_INNSATS("IKVAL"),
    SITUASJONSBESTEMT_INNSATS("BFORM"),
    BEHOV_FOR_ARBEIDSEVNEVURDERING("BKART");

    override fun fieldName(): String {
        return "innsatsgruppe"
    }

    override fun value(): String {
        return this.toString()
    }

    companion object {
        fun of(arenakode: String): Innsatsgruppe {
            return Arrays.stream(values())
                    .filter { innsatsgruppe: Innsatsgruppe -> innsatsgruppe.arenakode == arenakode }
                    .findAny().orElse(null)
        }

        @JvmStatic
        fun of(besvarelse: Besvarelse, alder: Int, harJobbetSammenhengendeSeksAvTolvSisteManeder: Boolean): Innsatsgruppe {
            return if (besvarelse.anbefalerBehovForArbeidsevnevurdering()) {
                BEHOV_FOR_ARBEIDSEVNEVURDERING
            } else if (besvarelse.anbefalerStandardInnsats(alder, harJobbetSammenhengendeSeksAvTolvSisteManeder)) {
                STANDARD_INNSATS
            } else {
                SITUASJONSBESTEMT_INNSATS
            }
        }
    }
}